### PR TITLE
Impl Patients and Receptionists

### PR DESCRIPTION
**Pull request by NikDub**
https://innowise-group.atlassian.net/browse/NIS-1436

### CODE REVIEW

File: ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            private readonly IPatientRepository _repository;
+            private readonly IMapper _mapper;
+
+            public AddPatientByPatientHandler(IPatientRepository repository, IMapper mapper)
+            {
+                _repository = repository;
+                _mapper = mapper;
+            }
+
+            public async Task<PatientDTO> Handle(AddPatientByPatientCommand request, CancellationToken cancellationToken)
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs]: If you pass cancellationToken, you need to add it to async methods

File: ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            {
+                if (request.patient == null)
+                    return null;
+
+                var patient = _mapper.Map<Patient>(request.patient);
+                patient.AccountId = request.AccountId;
+                patient.isLinkedToAccount = true;
+                await _repository.InsertAsync(patient);
+                await _repository.SaveAsync();
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs]: Let's not take SaveAsync to a separate method. Add it to all required methods inside the repo (insert, update, etc.). A person could call InsertAsync,  forget to call SaveAsync and then wonder why the entity hasn't been saved

File: ProfilesMicroService/Controllers/DoctorController.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+using Microsoft.AspNetCore.Mvc;
+using ProfilesMicroService.Application.DTO.Patient;
+using ProfilesMicroService.Application.Services.Abstractions;
+using ProfilesMicroService.Domain.Entities.Enums;
+
+namespace ProfilesMicroService.Api.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class DoctorController : Controller
```
Comment (kseniya-kalita) [inline ProfilesMicroService/Controllers/DoctorController.cs]: The common practise is to keep route names in plural.

File: ProfilesMicroService/Extensions/ServiceExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                                Type = ReferenceType.SecurityScheme,
+                                Id = "Bearer"
+                            },
+                            Name = "Bearer",
+                        },
+                        new List<string>()
+                    }
+                });
+
+                //setup.IncludeXmlComments(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Comments.xml"));
```
Comment (kseniya-kalita) [inline ProfilesMicroService/Extensions/ServiceExtensions.cs]: Remove or uncomment. We can't push commented changes (only if they are important Notes/ToDos fo the future)

File: ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+﻿using AutoMapper;
+using MediatR;
+using ProfilesMicroService.Application.DTO.Patient;
+using ProfilesMicroService.Application.Services.Abstractions;
+using ProfilesMicroService.Domain.Entities.Models;
+
+namespace ProfilesMicroService.Application.Services.CQRS.Commands
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs]: Let's keep the folder 'CQRS' at the same level as 'Services', not inside

File: ProfilesMicroService.Application/Services/PatientRepository.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
+﻿using Microsoft.EntityFrameworkCore;
+using ProfilesMicroService.Application.Services.Abstractions;
+using ProfilesMicroService.Domain.Entities.Models;
+using ProfilesMicroService.Infrastructure;
+
+namespace ProfilesMicroService.Application.Services
+{
+    public class PatientRepository : IPatientRepository
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/PatientRepository.cs]: Repository is not an Application layer.

File: ProfilesMicroService/Controllers/DoctorController.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            return Ok(doctor);
+        }
+
+        [HttpPost]
+        [Authorize(Roles = nameof(UserRole.Receptionist))]
+        public async Task<IActionResult> Create([FromBody] DoctorForCreateDTO model)
+        {
+            var doctor = await _doctorService.CreateAsync(model);
+            if (doctor == null)
+                return BadRequest();
```
Comment (kseniya-kalita) [inline ProfilesMicroService/Controllers/DoctorController.cs]: It's a good idead to put a comment on the reasons ob BadRequest/NotFound, etc.

File: ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public record AddPatientByPatientCommand(string AccountId, PatientForCreateDTO patient) : IRequest<PatientDTO>
+    {
+        public class AddPatientByPatientHandler : IRequestHandler<AddPatientByPatientCommand, PatientDTO>
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Commands/AddPatientByPatientCommand.cs]: Handler is stored in separated file. Command Is stored in sepaarated file. But in the same folder

File: ProfilesMicroService.Application/Services/DoctorService.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+using ProfilesMicroService.Application.Services.Abstractions;
+using ProfilesMicroService.Domain.Entities.Enums;
+using ProfilesMicroService.Domain.Entities.Models;
+using ProfilesMicroService.Infrastructure;
+
+namespace ProfilesMicroService.Application.Services
+{
+    public class DoctorService : IDoctorService
+    {
+        private readonly ApplicationDBContext _db;
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/DoctorService.cs]: Let's create a separate repo not to inject db context inside a service.

File: ProfilesMicroService.Infrastructure/Properties/launchSettings.json [hunk 10]
```diff
@@ -0,0 +1,12 @@
+{
+  "profiles": {
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Infrastructure/Properties/launchSettings.json]: I don't think that you need launchsettings in any project of type 'class library', only in Api

File: ProfilesMicroService.Infrastructure/ApplicationDBContext.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace ProfilesMicroService.Infrastructure
+{
+    public class ApplicationDBContext : DbContext
+    {
+        public DbSet<Doctor> Doctors { get; set; }
+        public DbSet<Receptionist> Receptionists { get; set; }
+        public DbSet<Patient> Patients { get; set; }
+
+        public DbSet<Status> Statuss { get; set; }
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Infrastructure/ApplicationDBContext.cs]: Misspelled, correct is 'Status'

File: ProfilesMicroService.Infrastructure/ApplicationDBContext.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+        public ApplicationDBContext(DbContextOptions<ApplicationDBContext> options) : base(options) { }
+
+        protected override void OnModelCreating(ModelBuilder builder)
+        {
+            base.OnModelCreating(builder);
+            builder.Entity<Receptionist>().Property(e => e.Id).HasDefaultValueSql("gen_random_uuid()");
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Infrastructure/ApplicationDBContext.cs:20]: I think it's redudant, Id of type Guid is automatically added as default behavior of ef

File: ProfilesMicroService.Application/Services/CQRS/Queries/GetPatientWithoutAccountQuery.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+            public GetPatientWithoutAccountHandler(IPatientRepository repository, IMapper mapper)
+            {
+                _repository = repository;
+                _mapper = mapper;
+            }
+
+            public async Task<IEnumerable<PatientDTO>> Handle(GetPatientWithoutAccountQuery request, CancellationToken cancellationToken)
+            {
+                var patientAll = await _repository.GetAllAsync();
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Queries/GetPatientWithoutAccountQuery.cs]: possible performance issues, you can have the whole method with where in repo

File: ProfilesMicroService.Application/Services/CQRS/Queries/GetPatientWithoutAccountQuery.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+                var patientList = patientAll.Where(r => !r.isLinkedToAccount)
+                    .Select(patient => isMatchPatient(patient, request.parient) ? patient : null);
+
+                return _mapper.Map<List<PatientDTO>>(patientList);
+            }
+
+            private bool isMatchPatient(Patient patient, PatientForCreateDTO requestPatient)
+            {
+                int countCoefficient = 0;
+                if (patient.FirstName.ToLower().Equals(requestPatient.FirstName.ToLower()))
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Queries/GetPatientWithoutAccountQuery.cs]: extract string comparisson in another method

File: ProfilesMicroService.Application/Services/CQRS/Commands/UpdatePatientCommand.cs [hunk 15]
```diff
@@ -0,0 +1,37 @@
+﻿using AutoMapper;
+using MediatR;
+using ProfilesMicroService.Application.DTO.Patient;
+using ProfilesMicroService.Application.Services.Abstractions;
+
+namespace ProfilesMicroService.Application.Services.CQRS.Commands
+{
+    public record UpdatePatientCommand(string id, PatientForUpdateDTO patient) : IRequest<PatientDTO>
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Commands/UpdatePatientCommand.cs]: properties in records from upper

File: ProfilesMicroService.Application/Services/CQRS/Queries/GetPatientWithoutAccountQuery.cs [hunk 16]
```diff
@@ -0,0 +1,50 @@
+﻿using AutoMapper;
+using MediatR;
+using ProfilesMicroService.Application.DTO.Patient;
+using ProfilesMicroService.Application.Services.Abstractions;
+using ProfilesMicroService.Domain.Entities.Models;
+
+namespace ProfilesMicroService.Application.Services.CQRS.Queries
+{
+    public record GetPatientWithoutAccountQuery(PatientForCreateDTO parient) : IRequest<IEnumerable<PatientDTO>>
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/Services/CQRS/Queries/GetPatientWithoutAccountQuery.cs]: misspelled

File: ProfilesMicroService.Infrastructure/ApplicationDBContext.cs (hunk 12 above)
Comment (NikDub) [reply ProfilesMicroService.Infrastructure/ApplicationDBContext.cs:20]: postgresql don't allow you to generate guid without this string

File: ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+                countCoefficient += 5;
+            if (isEquals(patient.MiddleName, requestPatient.MiddleName))
+                countCoefficient += 5;
+            if (patient.DateOfBirth.Equals(requestPatient.DateOfBirth))
+                countCoefficient += 3;
+
+            return countCoefficient >= 13;
+        }
+
+        private bool isEquals(string str1, string str2)
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs]: All methods from Upper

File: ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        public async Task<IEnumerable<PatientDTO>> Handle(GetPatientWithoutAccountQuery request, CancellationToken cancellationToken)
+        {
+            var patientList = (await _repository.GetAllAsync(cancellationToken)).Select(patient => isMatchPatient(patient, request.Patient) ? patient : null);
+            
+            return _mapper.Map<List<PatientDTO>>(patientList);
+        }
+
+        private bool isMatchPatient(Patient patient, PatientForCreateDTO requestPatient)
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs]: Same here

File: ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs [hunk 19]
```diff
@@ -0,0 +1,48 @@
+﻿using AutoMapper;
+using MediatR;
+using ProfilesMicroService.Application.DTO.Patient;
+using ProfilesMicroService.Domain.Entities.Models;
+using ProfilesMicroService.Infrastructure.Repository.Abstractions;
+
+namespace ProfilesMicroService.Application.CQRS.Queries
+{
+    public record GetPatientWithoutAccountQuery(PatientForCreateDTO Patient) : IRequest<IEnumerable<PatientDTO>>;
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs]: Let's separate such fiels into two: in one file we wave only query record, in another one handler

File: ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+        public GetPatientWithoutAccountHandler(IPatientRepository repository, IMapper mapper)
+        {
+            _repository = repository;
+            _mapper = mapper;
+        }
+
+        public async Task<IEnumerable<PatientDTO>> Handle(GetPatientWithoutAccountQuery request, CancellationToken cancellationToken)
+        {
+            var patientList = (await _repository.GetAllAsync(cancellationToken)).Select(patient => isMatchPatient(patient, request.Patient) ? patient : null);
```
Comment (kseniya-kalita) [inline ProfilesMicroService.Application/CQRS/Queries/GetPatientWithoutAccountQuery.cs]: Still performance issues, please think how to better do it

**Pull request merged**